### PR TITLE
Multiple build targets

### PR DESCRIPTION
**Pull request by sgomes**
Implement support for multiple build targets in Calypso.

We currently have a single build, which includes everything needed to support the lowest denominator in features across supported browsers. This results in a large amount of polyfills and syntactic transformations which aren't needed for the majority of users (who thankfully do use modern browsers), and ends up bloating the bundle and increasing browser download and execution times.

This PR aims to generate a separate evergreen build in production, and serve that build instead to evergreen browsers, by doing user agent inspection. Other browsers will still work, since they have the fallback bundle, but known modern browsers will get the benefit of a smaller bundle.

There's also a small "mustard test" that tries to prevent an old browser that somehow got the modern bundle from breaking (which shouldn't happen unless it's reporting a modern browser's UA). It does this by trying to determine in runtime if the browser "cuts the mustard" by checking for support for ES modules (a reasonable test for browser "modernness" at this point in time). If it fails, the browser gets redirected to a forced fallback build with `?forceFallback=1`. This is, as @blowery put it, a "belt and suspenders" approach, in trying to ensure that there is no breakage even if we somehow detect the UA incorrectly.

**Note:** This PR replaces #30420.

#### Changes proposed in this Pull Request

* Remove duplicate browserslist definition in Babel config.
* Use the browserslist definition to configure Terser ES6 and browser-specific settings.
* Add second browserslist definition to `package.json`, for evergreen browsers.
* Modify WebPack to output to different directories based on browserslist env.
* Modify npm scripts to run two builds in production (one for fallback, one for evergreen).
* Modify server-side code to read UA and choose a bundle accordingly.
* Add "mustard test" to HTML sent to evergreen browsers.

#### Testing instructions

Evergreen:
* Open in an evergreen browser (check browserslist definition for details of which browsers that means)
* Check that page resources are coming from `/calypso/evergreen/`
* Ensure that there is no breakage when using Calypso normally

Non-evergreen:
* Open in a non-evergreen but supported browser (e.g. IE 11).
* Check that page resources are coming from `/calypso/fallback` (not `/calypso/evergreen/`)
* Ensure that there is no breakage when using Calypso normally

Compare the two:
* Compare resource transfer sizes across both versions (particularly JS). The resources sent to the evergreen browser should be smaller and include ES6 syntax such as arrow functions (`=>`).

Forced fallback:
* Open any page with `?forceFallback=1` in an evergreen browser
* Check that page resources are coming from `/calypso/fallback` (not `/calypso/evergreen/`)
* Ensure that there is no breakage when using Calypso normally

### CODE REVIEW

File: client/document/index.jsx
Comment (blowery) [inline client/document/index.jsx]: can this just be `nomodule` without the property assignment?

**Comment by blowery** [comment]
This currently breaks `npm start` for dev work. It only builds the fallback build and the server bits that generate URLs don't detect that the evergreen build is unavailable.

It would be nice to just build the evergreen build and use that for dev work.

File: client/document/index.jsx
Comment (sgomes) [reply client/document/index.jsx]: The property assignment can be removed, but the capitalisation seems mandatory.

File: webpack.config.js
Comment (sgomes) [reply webpack.config.js]: Sounds good! Switched to `fallback`, as it seems like a more future-proof name.

**Comment by sgomes** [comment]
> This currently breaks `npm start` for dev work. It only builds the fallback build and the server bits that generate URLs don't detect that the evergreen build is unavailable.
> 
> It would be nice to just build the evergreen build and use that for dev work.

Great catch! I removed the `nomodule` safeguard from dev builds and switched them to evergreen. Hot reloading should still work correctly.

**Comment by sgomes** [comment]
Looks like the desktop build is failing due to dependency issues. Does anyone have any idea what the issue might be, there?

The Safari 10 test also seems to be failing with a timeout. Unclear if that's related to the PR, but taking a look at the other open PRs it seems that it might be, since the other ones don't appear to be failing.

File: package.json
Comment (simison) [inline package.json]: Would `last 2 versions, unreleased versions, not dead` rules cover these all at once?

https://github.com/browserslist/browserslist#full-list

File: webpack.config.js
Comment (simison) [inline webpack.config.js]: Could we add `isCalypsoClient &&`  check here not to write these when building using the SDK builder?

You can test with:
```
npm run sdk -- gutenberg client/gutenberg/extensions/presets/jetpack/ --output-dir=../somewhere-in-your-system --watch
```

https://wpcalypso.wordpress.com/devdocs/docs/sdk.md

File: webpack.config.js
Comment (simison) [inline webpack.config.js]: Could we have this be set to `defaults` if `! isCalypsoClient`?

Otherwise, SDK builder will always build `evergreen` and Jetpack and wpcom aren't ready for that just yet. :-)

If this can be a CLI argument in [`bin/sdk-cli.js`](https://github.com/Automattic/wp-calypso/blob/8dfcfced52ac061ff6cde1d25905da455cb1e5bb/bin/sdk-cli.js) that defaults to `defaults`, that would be ace!

File: package.json
Comment (sgomes) [reply package.json]: Unfortunately, that includes a bunch more browsers that are definitely not evergreen, such as IE11: https://browsersl.ist/?q=last+2+versions%2C+unreleased+versions%2C+not+dead

File: webpack.config.js
Comment (sgomes) [reply webpack.config.js]: Yup, not a problem, done! Is there something in this PR that causes this to become an issue?

File: webpack.config.js
Comment (sgomes) [reply webpack.config.js]: Seems like the WebPack build is doing more than I thought :-/ Done!

File: webpack.config.js
Comment (simison) [reply webpack.config.js]: Not at all and I recognise you didn't change this behaviour from how it was before — it's just unnecessary step in the SDK build and this seems like a good moment to add this check.

Potential annoyances could only be caused during development by running SDK build in watch mode while running Calypso, in which case assets from the SDK build could end up used in Calypso.

Comment (sgomes) [reply webpack.config.js]: Gotcha, happy to remove unnecessary work from the build! 👍

**Review by simison — APPROVED** [review]
From the SDK builder point of view, this is good now 👍

**Comment by jsnajdr** [comment]
One issue I found: the Gutenberg integration at `/block-editor/post/:site/:post` crashes when loading. The exception is swallowed in the async loader, but if you modify that code to report it, it turns out that `regeneratorRuntime` is missing.

The `@babel/polyfill` import no longer provides it, because the evergreen transpilation targets don't need it -- they natively support generators and async/await. But some Gutenberg packages (e.g., `@wordpress/api-fetch`) ship transpiled code that references the global. Filed an issue: WordPress/gutenberg#13890

File: babel.config.js
Comment (jsnajdr) [inline babel.config.js]: The `undefined` value means "read the browserlists config from a supported location", right? Worth a comment that explains it.

File: client/document/index.jsx
Comment (jsnajdr) [inline client/document/index.jsx]: Does this mean that fallback users who use IE11 et al will wee `forceFallback` in their URLs all the time?

File: package.json
Comment (jsnajdr) [inline package.json]: Do we use the `request` dependency anywhere?

**Comment by sgomes** [comment]
> One issue I found: the Gutenberg integration at `/block-editor/post/:site/:post` crashes when loading. The exception is [swallowed in the async loader](https://github.com/Automattic/wp-calypso/blob/3973f6ab16bbb560b6f4c04c6616e1e40cb474ab/client/components/async-loader/index.jsx#L32), but if you modify that code to report it, it turns out that `regeneratorRuntime` is missing.
> 
> The `@babel/polyfill` import no longer provides it, because the evergreen transpilation targets don't need it -- they natively support generators and async/await. But some Gutenberg packages (e.g., `@wordpress/api-fetch`) ship transpiled code that references the global. Filed an issue: [WordPress/gutenberg#13890](https://github.com/WordPress/gutenberg/issues/13890)

Thanks, @jsnajdr! Interesting issue. I wouldn't have expected the Gutenberg build to require something present that it isn't shipping with.

File: webpack.config.js
Comment (jsnajdr) [inline webpack.config.js]: Nice change! `AssetsWriter` is a Calypso specialty not needed anywhere else.

File: client/document/index.jsx
Comment (simison) [reply client/document/index.jsx]: Could the `forceFallback` be just initial way to do the switch and next page-loads would pick it up from `localStorage`?

Comment (sgomes) [reply client/document/index.jsx]: No, `forceFallback` will only be used if somehow we provide an evergreen bundle to a non-evergreen browser.

This is probably not super well explained in the PR, so I'll explain it here in more detail. The primary method of determining which build to use is the user agent string. If the user agent string for one of the known evergreen browsers is detected, then the browser gets sent the evergreen build. Otherwise it gets sent the fallback build.

This line here is a secondary mechanism designed to avoid breakage. It checks to see if the current browser supports ES modules (a good litmus test for browser "modernness" at this point in time), and if it doesn't, redirects the user. Now, this should only happen if the browser misreports its user agent string (e.g. by impersonating a newer browser), but since it leads to breakage, it made sense in my mind to add this escape hatch.

If we're happy relying on user agent strings alone, we can remove this check here and even remove `forceFallback` altogether (although it arguably could be left behind as a "hidden feature" for solving potential compatibility issues).

File: babel.config.js
Comment (sgomes) [reply babel.config.js]: Good point, added!

File: package.json
Comment (sgomes) [reply package.json]: The dependency (`useragent`) for one of our dependencies (`browserslist-useragent`) requires this to be specified if certain functionality is used, which our dependency has chosen to do. See https://www.npmjs.com/package/useragent#api

This was breaking the desktop build. I can try with just `yamlparser` and remove `request`, though.

File: client/document/index.jsx
Comment (jsnajdr) [reply client/document/index.jsx]: Thanks for the explanation. If it's just an emergency detection mechanism rather than the primary one, ugly URLs are OK and it's worth having it.

File: webpack.config.js
Comment (sgomes) [reply webpack.config.js]: Full credit goes to @simison on this one :)

File: package.json
Comment (sgomes) [reply package.json]: Submitted a commit to remove `request` as a direct dependency. We'll see how it goes!

Comment (jsnajdr) [reply package.json]: Hmm, so the `useragent` package optionally makes network requests during the build? I'd rather avoid that and have the build completely offline.

Until now, making network requests was an exclusive property of evil packages. E.g., `eslint-scope` that stealed your NPM credentials or `event-stream` that stealed your Bitcoins. Let's keep it that way 🙂

One day we might want to sandbox our builds and prevent them from making any network requests other than the NPM registry (and even that can be possibly cached).

If the `caniuse-db` package and the feature support database in `@babel/preset-env` can be offline, then a list of UA regexp can be offline, too.

Comment (sgomes) [reply package.json]: Those are some very good points.

I dig into this a bit deeper. As far as I can tell, there appears to be nothing in `browserslist-useragent` that would trigger a network request, since `useragent` isn't being initialised as the documentation link I provided states it should be.

My guess is that `yamlparser` is missing as an explicit dependency, and `request` is not being used at all. We'll know for sure once the current desktop build finishes.

Comment (sgomes) [reply package.json]: Confirmed, only `yamlparser` is needed. Build-time network requests avoided :)

**Comment by sgomes** [comment]
The PR now appears to build correctly for desktop. Safari 10 builds are still failing due to timeout issues, which is odd since Safari 10 should be seeing no changes (it should get the fallback build). No idea how I can investigate that further :-/

**Comment by blowery** [comment]
> No idea how I can investigate that further :-/

Browserstack can hit calypso.live (and local dev instances even) with Safari 10.

**Comment by sgomes** [comment]
> > No idea how I can investigate that further :-/
> 
> Browserstack can hit calypso.live (and local dev instances even) with Safari 10.

Thanks, @blowery ! Looks like `browserslist-useragent` can't handle parsing unreleased Safari versions very well. Not that 10 is unreleased, but since we're comparing it against our browserslist of Safari versions, and one of them is Safari TP, that breaks the library on any Safari version.

I've filed a bug (https://github.com/browserslist/browserslist-useragent/issues/38), but in the meantime I'm removing unreleased Safari versions from the `evergreen` list.

**Comment by sgomes** [comment]
The latest change seems to have fixed things! All tests currently pass, including desktop and Safari 10.

**Comment by sgomes** [comment]
Squashed it all into a single commit, to make it easier to keep up with rebases.

**Comment by blowery** [comment]
This looks awesome.

**Comment by blowery** [comment]
What's the scoop on https://github.com/WordPress/gutenberg/issues/13890? Are we still waiting on that to land and get integrated?

**Comment by sgomes** [comment]
@blowery: I believe that's still necessary for Gutenberg to work well inside Calypso, but as far as I know it's not needed for the `wp-admin` redirect? I'm not completely up to speed on the status there.

The live branch is https://calypso.live/?branch=try/multiple-build-targets , if you want to give it a try.

**Comment by sgomes** [comment]
@blowery Looking at the e2e tests, it looks like Gutenberg integration is failing, so we probably still need to wait on a release for the fix to WordPress/gutenberg#13890

**Comment by dmsnell** [comment]
> that's still necessary for Gutenberg to work well inside Calypso

we just merged #31416 and have switched to using an iframe for delivering the editor. you may want to rebase and try again to see if things are still broken

**Comment by sgomes** [comment]
Thanks, @dmsnell! 👍 Rebased, let's see how it goes.

**Comment by sgomes** [comment]
Rebased to pick up latest build process changes.

**Comment by sgomes** [comment]
Rebased to pick up build process changes from #31414.

**Comment by sgomes** [comment]
@blowery Looks like Gutenberg isn't an issue anymore, judging by the test results. Is there any way to make sure?

**Comment by blowery** [comment]
Maybe check with @Automattic/lannister, they're driving the removal of gutenberg bits from Calypso.

**Comment by gwwar** [comment]
Should be okay see https://github.com/Automattic/wp-calypso/pull/31416

File: babel.config.js
Comment (simison) [inline babel.config.js]: Since this PR has been open for a while and this line gets tossed around:

- https://github.com/Automattic/wp-calypso/pull/31464/files#diff-0e759b21864de1336af9c450cff5ad20R9
- https://github.com/Automattic/wp-calypso/pull/31536#issuecomment-474236981
- https://github.com/Automattic/wp-calypso/pull/31503/files#diff-f0bc8d1120b1704c18d6a492ba4ee38fR6

... it would be good to PR this change separately.

Comment (simison) [reply babel.config.js]: Landed in https://github.com/Automattic/wp-calypso/pull/31561

**Comment by sgomes** [comment]
Rebased to pick up the latest changes to the build process. In the `browserslist` config, `defaults` uses `@wordpress/browserslist-config`, with `evergreen` defined explicitly.

cc @simison @blowery

File: packages/calypso-build/webpack/minify.js
Comment (simison) [inline packages/calypso-build/webpack/minify.js]: Heads up @Automattic/gutenpack :-)

Changes in `packages/calypso-build/webpack/minify.js`

Comment (sgomes) [reply packages/calypso-build/webpack/minify.js]: Good call mentioning this, @simison. 👍 

To explain what's going on: the minify module will now try to configure Terser automatically based on the available `browserslist` config. Any explicitly provided `options` and `options.terserOptions` will be preserved, however, so this shouldn't cause any changes unless you're relying on defaults and those defaults are different from your browserslist config.

**Comment by sgomes** [comment]
I think all the pending issues around this PR have at this point been solved:
- All the tests in the full e2e suite pass
- Gutenberg shouldn't be a problem anymore
- ICFY can handle things nicely after the change now (it will look at the evergreen numbers)

The PR has been rebased and is a clean merge at time of this post.
The test branch is at https://calypso.live/?branch=try/multiple-build-targets .
The ICFY branch numbers are at http://iscalypsofastyet.com/branch?branch=try/multiple-build-targets (and they look quite nice! 😄) 

The PR is thus ready for a thorough final review! 🙂Once it's approved, I'd prefer to either hand over or coordinate release with Team Calypso, to ensure everything goes smoothly.

**Comment by sgomes** [comment]
Rebased after further changes to build system.

**Comment by blowery** [comment]
@sgomes Ran across this today: https://github.com/philipwalton/webpack-esnext-boilerplate/issues/1#issuecomment-445280691

Could you please check Edge and see what behavior it's exhibiting?

**Comment by sgomes** [comment]
> @sgomes Ran across this today: [philipwalton/webpack-esnext-boilerplate#1 (comment)](https://github.com/philipwalton/webpack-esnext-boilerplate/issues/1#issuecomment-445280691)
> 
> Could you please check Edge and see what behavior it's exhibiting?

Yes, I'll double-check.

I just want to point out that the `nomodule` thing we're doing is just a sanity check over an abundance of caution; the main mechanism is actually UA-based. We could easily remove the safeguard if it's causing issues.

**Comment by sgomes** [comment]
@blowery 

Edge 18 appears to be working correctly, loading resources a single time from the evergreen folder.

Edge 17 appears to be working correctly, loading resources a single time from the evergreen folder.

Edge 16 appears to be working correctly, loading resources a single time from the fallback folder.

Edge 15 has a hard fail ("This page is having a problem loading") loading `calypso.live` pages (not just the multiple builds branch), for some reason. It doesn't offer any useful errors.

**Comment by blowery** [comment]
@sgomes rockin! let's go!

**Review by blowery — APPROVED** [review]
LGTM

**Comment by dmsnell** [comment]
@sgomes after painful `git bisect`ing I tracked this PR to the introduction of one of the two `Critical dependency: the request of a dependency is an expression` errors I'm seeing in the `wp-desktop` build

```
WARNING in ./calypso/node_modules/browserslist/node.js 116:26-33
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./calypso/node_modules/browserslist/index.js
 @ ./calypso/node_modules/browserslist-useragent/lib/index.js
 @ ./calypso/server/pages/index.js
 @ ./calypso/server/boot/index.js
 @ ./desktop/server/server.js
 @ ./desktop/server/index.js
 @ ./desktop/app.js
 @ ./desktop/index.js

WARNING in ./calypso/node_modules/browserslist/node.js 116:18-66
Critical dependency: the request of a dependency is an expression
 @ ./calypso/node_modules/browserslist/index.js
 @ ./calypso/node_modules/browserslist-useragent/lib/index.js
 @ ./calypso/server/pages/index.js
 @ ./calypso/server/boot/index.js
 @ ./desktop/server/server.js
 @ ./desktop/server/index.js
 @ ./desktop/app.js
 @ ./desktop/index.js
```

When I open the build from this commit I get a blank electron app showing the following error:
```
Error: ENOENT: no such file or directory, open 'calypso/server/bundler/assets-evergreen.json'
    at Object.fs.openSync (fs.js:584:18)
    at Object.module.(anonymous function) [as openSync] (ELECTRON_ASAR.js:173:20)
    at Object.fs.readFileSync (fs.js:491:33)
    at Object.fs.readFileSync (ELECTRON_ASAR.js:505:29)
    at q (/Volumes/Optane/wp-desktop/release/mac/WordPress.com.app/Contents/Resources/app/build/desktop.js:319:2337)
    at Q (/Volumes/Optane/wp-desktop/release/mac/WordPress.com.app/Contents/Resources/app/build/desktop.js:319:3119)
    at X (/Volumes/Optane/wp-desktop/release/mac/WordPress.com.app/Contents/Resources/app/build/desktop.js:319:8954)
    at Layer.handle_error (/Volumes/Optane/wp-desktop/release/mac/WordPress.com.app/Contents/Resources/app/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/Volumes/Optane/wp-desktop/release/mac/WordPress.com.app/Contents/Resources/app/node_modules/express/lib/router/index.js:315:13)
    at /Volumes/Optane/wp-desktop/release/mac/WordPress.com.app/Contents/Resources/app/node_modules/express/lib/router/index.js:284:7
```

Any idea why we might be missing the assets file?

**Comment by sirreal** [comment]
> Any idea why we might be missing the assets file?

I spotted a suspect line in wp-desktop and opened a PR to see if it helps: https://github.com/Automattic/wp-desktop/pull/605

**Comment by sgomes** [comment]
> Any idea why we might be missing the assets file?

This is likely because the build process is generating one thing (the fallback build) while the server-side runtime is expecting another (the evergreen one). Likely a bug I introduced when I switched which build Desktop uses 😞

**Comment by sgomes** [comment]
> @sgomes after painful `git bisect`ing I tracked this PR to the introduction of one of the two `Critical dependency: the request of a dependency is an expression` errors I'm seeing in the `wp-desktop` build
> 
> ```
> WARNING in ./calypso/node_modules/browserslist/node.js 116:26-33
> Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
>  @ ./calypso/node_modules/browserslist/index.js
>  @ ./calypso/node_modules/browserslist-useragent/lib/index.js
>  @ ./calypso/server/pages/index.js
>  @ ./calypso/server/boot/index.js
>  @ ./desktop/server/server.js
>  @ ./desktop/server/index.js
>  @ ./desktop/app.js
>  @ ./desktop/index.js
> 
> WARNING in ./calypso/node_modules/browserslist/node.js 116:18-66
> Critical dependency: the request of a dependency is an expression
>  @ ./calypso/node_modules/browserslist/index.js
>  @ ./calypso/node_modules/browserslist-useragent/lib/index.js
>  @ ./calypso/server/pages/index.js
>  @ ./calypso/server/boot/index.js
>  @ ./desktop/server/server.js
>  @ ./desktop/server/index.js
>  @ ./desktop/app.js
>  @ ./desktop/index.js
> ```
> 

Hmm, these warnings are interesting. I have no idea what could be causing them, but they seem to becoming from inside the dependencies, right? Are they causing any failures?

**Comment by sirreal** [comment]
Related to the `Critical dependency: the request of a dependency is an expression`:
https://github.com/webpack/webpack/issues/196

**Comment by sirreal** [comment]
Also: https://github.com/browserslist/browserslist/issues/199

**Comment by sgomes** [comment]
> Also: [browserslist/browserslist#199](https://github.com/browserslist/browserslist/issues/199)

That's a much older version than what we're using. Ours should have the fix.

**Comment by sgomes** [comment]
In any case, the WebPack warning seems to be indicative of a potentially larger bundle, not any sort of failure. That's not an issue, since this is server-side code.

**Comment by sirreal** [comment]
The fix seems to be adding `browser` property to package.json:

```js
  "browser": {
    "./node.js": "./browser.js",
    "path": false
  },
```
[source](https://github.com/browserslist/browserslist/blob/68f270828f74d4209ddfea75b4e12429163a1d9f/package.json#L46-L49)

According to the documentation ([#](https://webpack.js.org/configuration/resolve/#resolvealiasfields), [#](https://github.com/defunctzombie/package-browser-field-spec#replace-specific-files---advanced)), that would mean "`node.js` should be replaced by `browser.js`. That doesn't seem to be happening according to the stack trace on those warnings 😕

**Comment by sgomes** [comment]
It's not clear to me that's the right fix for us. That sounds like a fix for code that runs in a browser environment, whereas ours runs in a node environment.

**Comment by dmsnell** [comment]
> In any case, the WebPack warning seems to be indicative of a potentially larger bundle, not any sort of failure. That's not an issue, since this is server-side code.

As far as I saw it didn't cause errors but I didn't check all the corners of the app. After going back and re-bisecting it looks like the `iconv` warning was there before this PR but the `browserlist` one is introduced here.

Thanks everyone for your efforts to clean this up. Since the desktop app is building again I think that I was partially mislead by the warnings. The assets are now working and the app runs at least…

**Pull request merged**
